### PR TITLE
Test Windows with GitHub Actions

### PR DESCRIPTION
**Pull request by nulano**
For #3606 

Adds Windows tests to GitHub Actions.

Existing scripts `winbuild\build_dep.py` and `winbuild\build.py` expect MSVS2015, but GitHub Actions only comes with MSVS2017 (or MSVS2019). It does come with a backwards-compatibility component, but it is incompatible with the existing scripts. For now this uses build scripts hardcoded in `test-windows.yml`, I would like to update the `winbuild` scripts once I have all the dependencies working.

Dependency builds are split into separate steps to make logs easier to read. If kept, this would require another change in `winbuild\build_dep.py`.

## Available dependencies:

* [x] libjpeg
* [x] libjpeg-turbo
* [x] zlib
* [x] webp
* [x] libtiff
* [x] freetype
* [x] lcms2
* [x] openjpeg
* [x] ghostscript
* [x] ~libimagequant~
* [x] ~raqm~

_Strikethrough indicates potential licensing issues for binary distribution._

## TODO

* [x] Only Python 3.6.8 for now. TODO add 3.5.4 and 3.7.4
* [x] Only x86 for now. TODO add x64
* [x] Add PyPy3
* [x] Make sure dependencies are built for Release
* [x] Fix all LNK4098 warnings
* [ ] ~Update `winbuild\build_dep.py` and `winbuild\build.py`~ (in future PR)
* [x] Upload coverage. Depends on #4081 for token.
* [x] Build wheels

### CODE REVIEW

File: .github/workflows/test-windows.yml [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        echo on
+        rmdir /S /Q output\release-static
+        nmake -nologo -f Makefile.vc CFG=release-static RTLIBCFG=static OBJDIR=output ARCH=x86 all
+        mkdir %INCLIB%\webp
+        copy /Y /B src\webp\*.h %INCLIB%\webp
+        copy /Y /B output\release-static\x86\lib\* %INCLIB%
+
+    - name: Build dependencies / freetype
+      run: |
+        REM Toolkit v100 not available; missing VCTargetsPath; Clean failes
```
Comment (radarhere) [inline .github/workflows/test-windows.yml]:
```suggestion
        REM Toolkit v100 not available; missing VCTargetsPath; Clean fails
```

Comment (nulano) [reply .github/workflows/test-windows.yml]: Fixed

File: .github/workflows/test-windows.yml [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        nmake -nologo -f Makefile clean
+        nmake -nologo -f Makefile
+        mkdir %INCLIB%\openjpeg-2.3.1
+        copy /Y /B src\lib\openjp2\*.h %INCLIB%\openjpeg-2.3.1
+        copy /Y /B bin\*.lib %INCLIB%
+
+    - name: Build dependencies / ghostscript
+      if: false
+      run: |
+        REM only used for Python 2.7 in AppVeyor?
```
Comment (radarhere) [inline .github/workflows/test-windows.yml]: What makes you think that we're currently only using Ghostscript in Python 2.7?

Comment (nulano) [reply .github/workflows/test-windows.yml]: https://github.com/python-pillow/Pillow/blob/1608a15ab5e18b96e0145e66c5218e4adee3fbc0/winbuild/build_dep.py#L323
But looking at AppVeyor logs, GhostScript is available in tests.

Comment (nulano) [reply .github/workflows/test-windows.yml]: I figured it out: `C:\Python27\` is used because it is included in `%PATH%`.

**Comment by hugovk** [comment]
Great work!

The 6 * Python 3 jobs take ~28 minutes on AppVeyor (in serial), and ~7 minutes on Actions (all in parallel)!

Don't worry about 2.7, we're dropping it soon.

---

Perhaps `exclude` can be used to test PyPy3 x86 and exclude PyPy3 x64, something like this?

```yml
runs-on: ${{ matrix.os }}
strategy:
  matrix:
    os: [macOS-10.14, windows-2016, ubuntu-18.04]
    node: [4, 6, 8, 10]
    exclude:
      # excludes node 4 on macOS
      - os: macos-10.14
        node: 4
```

Example from https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix

---

I'm generally fine with refactoring `winbuild`. We're not using AppVeyor to release anything, a lack of dependencies was the main blocker for that: https://github.com/python-pillow/Pillow/issues/553#issuecomment-488551254. So right now, it won't break any Windows releases.

It would be a great longer term goal to build Windows wheels via CI, to remove the [one-person bus factor](https://github.com/python-pillow/Pillow/blob/master/RELEASING.md#windows). 

And if we do so, I'd favour doing so with Actions so we can retire or at least minimise AppVeyor.

File: .github/workflows/test-windows.yml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+jobs:
+  build:
+  
+    runs-on: windows-2016
+    strategy:
+      matrix:
+        python-version:
+          - "3.5.4"
+          - "3.6.8"
+          - "3.7.4"
```
Comment (hugovk) [inline .github/workflows/test-windows.yml]: Do we need to specify the full 3.x.y version, or can we use 3.x to have the latest one available, like for Linux and macOS?

https://github.com/python-pillow/Pillow/pull/4081/files#diff-fe8421955fd596131bb6f1b78984b2fbR17

Comment (nulano) [reply .github/workflows/test-windows.yml]: Fixed on my branch, thanks!

**Comment by nulano** [comment]
> Perhaps `exclude` can be used to test PyPy3 x86 and exclude PyPy3 x64, something like this?
> [...]
> Example from https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix

Done on my branch, thank you!

I have 21 hung jobs I can't cancel due to #4074, once they time out and I can test it, I will push it here and mark Ready for review.

-----

> I'm generally fine with refactoring `winbuild`. We're not using AppVeyor to release anything, a lack of dependencies was the main blocker for that: [#553 (comment)](https://github.com/python-pillow/Pillow/issues/553#issuecomment-488551254). So right now, it won't break any Windows releases.
> 
> It would be a great longer term goal to build Windows wheels via CI, to remove the [one-person bus factor](https://github.com/python-pillow/Pillow/blob/master/RELEASING.md#windows).

https://github.com/python-pillow/Pillow/pull/4029#issuecomment-522343593:
> Those files are produced by my build script, which simply runs **setup.py** with bdist_wheel, bdist_egg, and bdist_wininst arguments for all supported CPython versions.

I remembered this as **build.py**... I'll go ahead and refactor `winbuild` once the rest of this PR is done.

----

> And if we do so, I'd favour doing so with Actions so we can retire or at least minimise AppVeyor.

Agreed.

**Comment by hugovk** [comment]
> I have 21 hung jobs I can't cancel due to #4074, once they time out and I can test it, I will push it here and mark Ready for review.

Can you click the "Cancel check suite" button?

https://help.github.com/en/articles/managing-a-workflow-run#canceling-a-workflow-run

**Comment by nulano** [comment]
> Can you click the "Cancel check suite" button?

It gives me an error.
I reported the issue here: https://github.community/t5/GitHub-Actions/Can-t-cancel-hung-job-quot-Failed-to-cancel-check-suite-quot/m-p/32673#M1216

image

**Comment by hugovk** [comment]
6 hour timeout, yikes.

I've pushed your `gha-win-wip` branch to my fork to run it here:

https://github.com/hugovk/Pillow/commit/9def3c588a1dd56f32338c628552dce77a6c10b4/checks

All green except PyPy3.

---

I saw the check for the `GITHUB_ACTIONS` env var here:

https://github.com/nulano/Pillow/commit/9def3c588a1dd56f32338c628552dce77a6c10b4#diff-2deb7e4843eb6ebdbc14446813affc8eR359

I've not tested it, but it looks like it should be `GITHUB_ACTION` without the `S`:

* https://github.community/t5/GitHub-Actions/Have-the-CI-environment-variable-set-by-default/m-p/32671/highlight/true#M1215
* https://help.github.com/en/articles/virtual-environments-for-github-actions#default-environment-variables

**Comment by nulano** [comment]
> 6 hour timeout, yikes.
> 
> I've pushed your `gha-win-wip` branch to my fork to run it here:

Thanks!

----

> https://github.com/hugovk/Pillow/commit/9def3c588a1dd56f32338c628552dce77a6c10b4/checks
> 
> All green except PyPy3.

    debug: OperationError:
    debug:  operror-type: LookupError
    debug:  operror-value: cp65001 encoding is only available on Windows

Looks like yet another PyPy bug, sigh...

I'll confirm this tomorrow by skipping the failing test.

----

> I've not tested it, but it looks like it should be `GITHUB_ACTION` without the `S`:

Hmm... I've run `cmd /c "set"` to dump all enviroment variables and there is an undocumented variable `GITHUB_ACTIONS=true`: https://github.com/nulano/Pillow/commit/4b4917a31c69b2a0deb143b01b5feb839a83b937/checks#step:2:25

I suppose `GITHUB_ACTION` will suffice until the docs are updated.

**Comment by hugovk** [comment]
GitHub staff replied:

> There's two distinct values here.  GITHUB_ACTION is the name of the action that's running, if any (and there may not be one).  GITHUB_ACTIONS is always set to show that you're running within GitHub Actions.

https://github.community/t5/GitHub-Actions/Have-the-CI-environment-variable-set-by-default/td-p/32125

Let's stick with the plural.

**Comment by nulano** [comment]
PyPy3 added, marking ready for review. The remaining tasks can be done in future PRs if this is merged in the meantime.

I added a 30min timeout (triple PyPy3 build time) to avoid what happened yesterday.

----

PyPy bug reported and test skipped: https://bitbucket.org/pypy/pypy/issues/3081/operationerror-cp65001-is-only-available

File: Tests/test_main.py [hunk 4]
```diff
@@ -5,8 +5,16 @@
 import sys
 from unittest import TestCase
 
+from .helper import on_github_actions, unittest
+
 
 class TestMain(TestCase):
+    @unittest.skipIf(
+        sys.platform == "win32"
+        and hasattr(sys, "pypy_translation_info")
```
Comment (hugovk) [inline Tests/test_main.py]: Minor thing, there's an `is_pypy()` in `test_core_resources.py` that could be moved to `helper.py`

File: Tests/test_imageshow.py [hunk 5]
```diff
@@ -34,7 +36,10 @@ def show_image(self, image, **options):
         # Restore original state
         ImageShow._viewers.pop(0)
 
-    @unittest.skipUnless(on_ci(), "Only run on CIs")
+    @unittest.skipUnless(
+        on_ci() and not (sys.platform == "win32" and on_github_actions()),
+        "Only run on CIs; hangs on Windows on Github Actions",
```
Comment (hugovk) [inline Tests/test_imageshow.py]:
```suggestion
        "Only run on CIs; hangs on Windows on GiHhub Actions",
```

File: .github/workflows/test-windows.yml [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        echo on
+        rmdir /S /Q Lib
+        rmdir /S /Q Projects\VC2015\Release
+        set VCTargetsPath=C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets
+        set MSBUILD="C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\MSBuild.exe"
+        %MSBUILD% Projects\VC2015\lcms2.sln /t:Clean;lcms2_static /p:Configuration="Release" /p:Platform=${{ matrix.platform-msbuild }} /m
+        xcopy /Y /E /Q include %INCLIB%
+        copy /Y /B Lib\MS\*.lib %INCLIB%
+
+    - name: Build dependencies / openjpeg
```
Comment (hugovk) [inline .github/workflows/test-windows.yml]:
```suggestion
    - name: Build dependencies / OpenJPEG
```

File: .github/workflows/test-windows.yml [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        rmdir /S /Q objs
+        set DefaultPlatformToolset=v140
+        set VCTargetsPath=C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets
+        set MSBUILD="C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\MSBuild.exe"
+        %MSBUILD% builds\windows\vc2010\freetype.sln /t:Build /p:Configuration="Release" /p:Platform=${{ matrix.platform-msbuild }} /m
+        xcopy /Y /E /Q include %INCLIB%
+        copy /Y /B objs\${{ matrix.platform-msbuild }}\Release\freetype.dll %INCLIB%
+        copy /Y /B objs\${{ matrix.platform-msbuild }}\Release\freetype.lib %INCLIB%
+
+    - name: Build dependencies / lcms2
```
Comment (hugovk) [inline .github/workflows/test-windows.yml]:
```suggestion
    - name: Build dependencies / LCMS2
```

File: .github/workflows/test-windows.yml [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        cd /D %BUILD%\libwebp-1.0.3
+        call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" ${{ matrix.platform-vcvars }} 8.1
+        echo on
+        rmdir /S /Q output\release-static
+        nmake -nologo -f Makefile.vc CFG=release-static RTLIBCFG=static OBJDIR=output ARCH=${{ matrix.architecture }} all
+        mkdir %INCLIB%\webp
+        copy /Y /B src\webp\*.h %INCLIB%\webp
+        copy /Y /B output\release-static\${{ matrix.architecture }}\lib\* %INCLIB%
+
+    - name: Build dependencies / freetype
```
Comment (hugovk) [inline .github/workflows/test-windows.yml]:
```suggestion
    - name: Build dependencies / FreeType
```

File: .github/workflows/test-windows.yml [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" ${{ matrix.platform-vcvars }} 8.1
+        echo on
+        copy %GITHUB_WORKSPACE%\winbuild\nmake.opt nmake.opt
+        nmake -nologo -f makefile.vc clean
+        nmake -nologo -f makefile.vc lib
+        copy /Y /B libtiff\tiff*.h %INCLIB%
+        copy /Y /B libtiff\*.dll %INCLIB%
+        copy /Y /B libtiff\*.lib %INCLIB%
+
+    - name: Build dependencies / webp
```
Comment (hugovk) [inline .github/workflows/test-windows.yml]:
```suggestion
    - name: Build dependencies / WebP
```

File: .github/workflows/test-windows.yml [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        cd /D %BUILD%\zlib-1.2.11
+        call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" ${{ matrix.platform-vcvars }} 8.1
+        echo on
+        nmake -nologo -f win32\Makefile.msc clean
+        nmake -nologo -f win32\Makefile.msc zlib.lib
+        copy /Y /B z*.h %INCLIB%
+        copy /Y /B *.lib %INCLIB%
+        copy /Y /B zlib.lib %INCLIB%\z.lib
+
+    - name: Build dependencies / libtiff
```
Comment (hugovk) [inline .github/workflows/test-windows.yml]:
```suggestion
    - name: Build dependencies / LibTIFF
```

File: Tests/test_main.py (hunk 4 above)
Comment (nulano) [reply Tests/test_main.py]: How about moving `sys.platform == "win32"` to `helper.is_win32()` as well? A quick grep found 20 matches in `Tests`.

Comment (hugovk) [reply Tests/test_main.py]: Sure! There's also a `skipKnownBadTest` in `helper.py`. This could be done in another PR if you prefer.

Comment (nulano) [reply Tests/test_main.py]: `is_pypy` and `is_win32` are done, let's do `skipKnowBadTest` in a future PR.

**Comment by nulano** [comment]
> Some minor things

Done.

----

> It would be a great longer term goal to build Windows wheels via CI, to remove the one-person bus factor.
>
> And if we do so, I'd favour doing so with Actions so we can retire or at least minimise AppVeyor.

Added wheel building. Note that these are not release-ready for now (e.g. `freetype.dll` is not added to wheels).

Skipping PyPy for now, as it was overwriting CPython eggs.

Set to run only on push, as the [file size limit is not clear](https://github.com/actions/upload-artifact/issues/9). It should be possible to download wheels from PR author's push.

**Comment by nulano** [comment]
LCMS2 is now working (both AppVeyor and GitHub Actions) and FreeType is linked statically (only GitHub Actions, AppVeyor failed to link).

Comparing the generated wheel with the 6.1.0 release on my PC I get:

<pre>
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image, features
>>> Image.__version__
<b>'6.1.0'</b>
>>> features.get_supported_modules()
<b>['pil', 'tkinter', 'freetype2', 'littlecms2', 'webp']</b>
>>> features.get_supported_features()
<b>['webp_anim', 'webp_mux', 'transp_webp', 'libjpeg_turbo']</b>
>>> features.get_supported_codecs()
<b>['jpg', 'jpg_2000', 'zlib', 'libtiff']</b>
</pre>

<pre>
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image, features
>>> Image.__version__
<b>'6.2.0.dev0'</b>
>>> features.get_supported_modules()
<b>['pil', 'tkinter', 'freetype2', 'littlecms2', 'webp']</b>
>>> features.get_supported_features()
<b>['webp_anim', 'webp_mux', 'transp_webp']</b>
>>> features.get_supported_codecs()
<b>['jpg', 'jpg_2000', 'zlib', 'libtiff']</b>
</pre>

It looks like libjpeg_turbo is now the only missing dependency on GitHub Actions.

I have also found instructions to build raqm on Windows: https://github.com/HOST-Oman/libraqm-cmake#windows-visualstudio-etc
Is there any reason raqm is not included in Windows releases?

**Comment by hugovk** [comment]
> I have also found instructions to build raqm on Windows: [HOST-Oman/libraqm-cmake#windows-visualstudio-etc](https://github.com/HOST-Oman/libraqm-cmake#windows-visualstudio-etc)
> Is there any reason raqm is not included in Windows releases?

It's probably because Windows setup wasn't ready when we first added it, but could be some licensing issue, I can't remember. Or am I mixing up with some other library?

**Comment by radarhere** [comment]
https://pillow.readthedocs.io/en/stable/installation.html#external-libraries
> Windows support: Raqm support is currently unsupported on Windows.

**Comment by wiredfool** [comment]
Initial release was source only, because of license incompatibility. Then we shifted to being able to dynamically load the library if it was available (but still not distributing it).  I don’t remember if that ever was tested on windows, but I know that the library was one of the more difficult ones for me to build on windows when I was working in it.

**Comment by wiredfool** [comment]
@radarhere.  That note sounds like me giving up on trying to build it for windows.

**Comment by wiredfool** [comment]
Also, @nulano I just want to thank you for tackling the windows builds.  They weren’t much fun to set up in the first place, but they’re really important for a segment of our users.

**Comment by radarhere** [comment]
This is impressive. It's something that I've struggled with, and you're blazing through it.

libimagequant is another item for your checklist - https://github.com/python-pillow/Pillow/pull/4084/checks?check_run_id=237231059#step:15:417

image

**Comment by nulano** [comment]
> > Is there any reason raqm is not included in Windows releases?
> 
> It's probably because Windows setup wasn't ready when we first added it, but could be some licensing issue, I can't remember. Or am I mixing up with some other library?

From https://github.com/python-pillow/Pillow/issues/2567#issuecomment-312439164:
> Built without RAQM, which seems to be GPL contaminated.

Seems to be from FriBidi, which is licensed under LGPL 2.1 or later: https://github.com/fribidi/fribidi#license

----

> libimagequant is another item for your checklist - https://github.com/python-pillow/Pillow/pull/4084/checks?check_run_id=237231059#step:15:417

I'll also add it to `features.py`.
Note that it was relicensed under GPL 3 (https://github.com/ImageOptim/libimagequant/commit/d7c9ecae0492bbdd449fd8e3ccf94abf73ea5f36#diff-97d9303acdfc078a050e61dc5c1a9a76) and is indeed not present in the 6.1.0 release:

```
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image
>>> Image.__version__
'6.1.0'
>>> Image.new("RGB",(50,50)).quantize(100,Image.LIBIMAGEQUANT)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "F:\Languages\Python37x64\lib\site-packages\PIL\Image.py", line 1128, in quantize
    im = self._new(self.im.quantize(colors, method, kmeans))
ValueError: dependency required by this method was not enabled at compile time
```

For now, I have been focusing on getting this as close to release-able as I can. I think there are only two issues with releasing using this right now: 1) missing `libjpeg-turbo` 2) test the wheels (compare performance with previous release, i.e. is everything compiled with proper optimization?; and check there are no system dependency issues, i.e. does it work on a clean Vista install?)

**Comment by cgohlke** [comment]
> Windows support: Raqm support is currently unsupported on Windows.

>  I don’t remember if that ever was tested on windows

FWIW, all tests pass when a proper `libraqm.dll` is present.

**Comment by nulano** [comment]
I now have Raqm builds working in my branch; CPython passes all tests.

In PyPy3 this has introduced a seemingly unrelated failure in `test_imagedraw.test_textsize_stroke()` from #3978: https://github.com/nulano/Pillow/runs/241395693#step:21:1065

Any ideas @radarhere ?

**Comment by radarhere** [comment]
Nothing particularly helpful, no - I do wonder if the PyPy3 nightly would also trigger the problem on Linux.

**Comment by nulano** [comment]
Actually, all freetype tests are broken on PyPy3 with Raqm, so it's probably my fault...

For example:

a

**Comment by hugovk** [comment]
I'd really like to get this PR merged in soon.

It'd be fine for PyPy3 to follow in a later PR, the main thing would be to replace the bulk of the AppVeyor jobs. What is important before removing AppVeyor jobs is coverage.

Having said that, there's a lot of value in having these GHA jobs merged in and running in parallel with AppVeyor, even if coverage is to follow, just because they're so fast and we'll get build feedback much quicker.

---

Coverage:

> * [x]  Upload coverage. Depends on #4081 for token.

I added a token to `.codecov.yml` in #4081. I'm not yet sure if the coverage uploaded to Codecov on my fork and the Pillow repo are fully consistent, but feel free to grab the same token to try out here too. 

Another place it go is hardcoded in test*.yml:
```yml
      env:
        CODECOV_TOKEN: abcde12345uvwxyz
```

Ideally forks could still have correct coverage showing on their own Codecov page, before PRs are made. And then a PR would have correct coverage on the Pillow Codecov page.

Thanks!

**Comment by aclark4life** [comment]
@hugovk Merge-away from me pending everyone else's comfort level and everyone please allow me to echo @wiredfool thanks to @nulano !! (@nulano you should consider joining the Pillow team if you are interested cf. #4048 . We could add you to the team now as a volunteer, then if that goes well, add you to Tidelift in 2020 …)

**Comment by cgohlke** [comment]
Is there a reason why the external C libraries are built/linking against the Multi-threaded (`/MT`) runtime library instead of Multi-threaded DLL (`/MD`), which is used by Python/distutils?

**Comment by nulano** [comment]
> Is there a reason why the external C libraries are built/linking against the Multi-threaded (`/MT`) runtime library instead of Multi-threaded DLL (`/MD`), which is used by Python/distutils?

@cgohlke Look at the Build Pillow step. I noticed that distutils chooses /MT for CPython 3.x, but /MD for PyPy3. I'm not sure whether there was a change in CPython, but I guessed that there was based on this. I do remember seeing /MD in Python 2.7 builds on AppVeyor.

--- 

> I'd really like to get this PR merged in soon.
>
> It'd be fine for PyPy3 to follow in a later PR, the main thing would be to replace the bulk of the AppVeyor jobs. What is important before removing AppVeyor jobs is coverage.
>
> Having said that, there's a lot of value in having these GHA jobs merged in and running in parallel with AppVeyor, even if coverage is to follow, just because they're so fast and we'll get build feedback much quicker.

I pushed Raqm with PyPy3 disabled for now. My current theory is that this is due to the /MD vs /MT difference between CPython and PyPy, but I think this can be tested later.

I would definitely make sure AppVeyor is kept functional, at least until GHA is out of beta; just reducing it to one or two jobs would probably be a good start. This is a good idea especially when you consider yesterday's 2 hour outage: https://github.com/actions/checkout/issues/49

I would be more comfortable adding in the token in a separate PR, so feel free to merge now.

---

> [...] (@nulano you should consider joining the Pillow team if you are interested cf. #4048 . We could add you to the team now as a volunteer, then if that goes well, add you to Tidelift in 2020 …)

I will definitely consider this, but I expect to become quite busy over the next few weeks, so I'll get back to you later once I've got my schedule figured out.

**Comment by cgohlke** [comment]
> I noticed that distutils chooses /MT for CPython 3.x

That should not be. I would recommend not to distribute these builds. Do the Visual Studio installations include the redistributable files? Could be that distutils fails to find the vcruntime_redist files and switches to /MT.

**Comment by nulano** [comment]
> > I noticed that distutils chooses /MT for CPython 3.x
> 
> That should not be. I would recommend not to distribute these builds. Do the Visual Studio installations include the redistributable files? Could be that distutils fails to find the vcruntime_redist files and switches to /MT.

Looking at the source of distutils on my desktop, this does seem to make sense. I am pretty sure the Visual Studio install does have the required files, both because the PyPy3 distutils finds it, and because I see the file on my desktop Visual Studio 2017 Community install (in `C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Redist\MSVC\14.14.26405\x86\Microsoft.VC141.CRT\vcruntime140.dll`). I'm not sure why CPython's distutils can't see it...

**Comment by nulano** [comment]
I found the issue. After removing this line, Pillow is built with /MD: https://github.com/python-pillow/Pillow/blob/ae8fd7446f9520aed3cec75d04caa9ff94f822ce/.github/workflows/test-windows.yml#L316

Looking through _msvccompiler setting this value makes distutils skip looking for the redistributable files and `env['py_vcruntime_redist']` is never set. Does anyone remember why this line was added? I only added it because I copied it from AppVeyor.

**Comment by aclark4life** [comment]
@nulano Probably fine pending @cgohlke's confirmation … FWIW, here's some docs about that environment variable: https://docs.python.org/3/distutils/apiref.html#module-distutils.msvccompiler

**Comment by cgohlke** [comment]
> set DISTUTILS_USE_SDK=1 

That is no longer needed. Probably a leftover from the times when Visual Studio 2008 and 2010 were not free (or missing 64-bit compilers) while the Windows SDKs were.

**Comment by nulano** [comment]
Everything is /MD now (except ghostscript, which is only called through the exe). PyPy3 still fails with Raqm, might be a PyPy bug.

I think the AppVeyor failure is intermittent, I faintly remember a similar problem when I was working on PyPy3 for AppVeyor. Restarting it might fix it.

I also ran all tests on a clean Vista VM, getting two failures: missing Visual Studio and missing PowerShell, both only needed for Tests.

**Comment by hugovk** [comment]
Are we good to merge this once the conflicts are resolved?  

> I would definitely make sure AppVeyor is kept functional, at least until GHA is out of beta; just reducing it to one or two jobs would probably be a good start. This is a good idea especially when you consider yesterday's 2 hour outrage: [actions/checkout#49](https://github.com/actions/checkout/issues/49)

Good point, let's keep both ticking over until at least after launch. We can use that time for any updates/tuning needed for this.

**Comment by hugovk** [comment]
Thank you very much!

**Comment by tvatter** [comment]
Quick question, I managed to follow some of your steps, but only using `windows-2016` and not `windows-latest`. Given that it's going to be deprecated tomorrow, it's kind of a bummer. Have you thought about this already?

**Comment by radarhere** [comment]
@tvatter see #4188

**Comment by tvatter** [comment]
Thanks! (what a mess ^^)

**Pull request merged**